src/Huntress/Plugin/Stonks.php

```diff
@@ -0,0 +1,155 @@
+<?php
+
+
+namespace Huntress\Plugin;
+
+
+use CharlotteDunois\Yasmin\Models\MessageEmbed;
+use Huntress\EventData;
+use Huntress\EventListener;
+use Huntress\Huntress;
+use Huntress\PluginHelperTrait;
+use Huntress\PluginInterface;
+use Scheb\YahooFinanceApi\ApiClient;
+use Scheb\YahooFinanceApi\ApiClientFactory;
+use Scheb\YahooFinanceApi\Results\Quote;
+
+/**
+ * Load ticker data from the Yahoo Finance API.
+ *
+ * @author Aran <aran@ermarian.net>
+ */
+class Stonks implements PluginInterface
+{
+
+    use PluginHelperTrait;
+
+    public const CACHE_TTL = 300;
+
+    /**
+     * @var \Scheb\YahooFinanceApi\ApiClient
+     */
+    private ApiClient $client;
+
+    /** @var array[] */
+    private array $cache;
+
+    /** @var int */
+    private int $ttl;
+
+    public function __construct()
+    {
+        $this->client = ApiClientFactory::createApiClient();
+        $this->cache = [];
+        $this->ttl = static::CACHE_TTL;
+    }
+
+    public static function register(Huntress $bot): void
+    {
+        $instance = new static();
+        $bot->eventManager->addEventListener(
+            EventListener::new()
+                ->addCommand('stonks')
+                ->setCallback([$instance, 'stonks'])
+        );
+    }
+
+    public function stonks(EventData $event)
+    {
+        $arg = explode(' ', strtoupper(self::arg_substr($event->message->content, 1) ?? ''));
+        $channel = $event->message->channel;
+        $now = time();
+        $get = [];
+        // Check which quotes must be refreshed.
+        foreach ($arg as $symbol) {
+            if (!isset($this->cache[$symbol]) || $this->cache[$symbol]['time'] + $this->ttl < $now) {
+                $get[] = $symbol;
+            }
+        }
+        // Get new data.
+        if ($get) {
+            $data = $this->client->getQuotes($get);
```

Comment:
looking at the deps, this appears to use guzzle on the back end i think? which is woefully blocking. If the library can support it please switch to using non-blocking network stuff (perhaps with [`URLHelpers::makeRequest()`](https://sylae.github.io/Yasmin/master/CharlotteDunois/Yasmin/Utils/URLHelpers.html#method_makeRequest)?), or alternatively use React's ChildProcess functionality to run it in a separate thread. There's a janky example of how to do that [here](https://github.com/sylae/huntress/blob/7b0c6cbdaa01cf1cf9b1a87e1a80d8b258440631/src/Huntress/Plugin/WormRP.php#L381-L428).

unfortunately with the main Huntress instance running on as many guilds and doing as many things as it does, we've gotta be really aware of not using blocking stuff when we can (in fact, one of the 😬 things im putting off is switching everything i have to using non-blocking filesystem requests)

Reply:
well, that wasn't easy, but I think we're there now :) https://github.com/arancaytar/yahoo-finance-api